Friendlier error messages

dirkgr opened the pull request:
Fixes #21, and similar things.

This gives the datastore API slightly richer error messages, and the datastore CLI much more friendly output.

@schmmd, FYI

jkinkead commented on datastore-cli/src/main/scala/org/allenai/datastore/cli/Common.scala:
```diff
@@ -1,9 +1,23 @@
     System.err.println("To specify another datastore (such as 'private'), use `-d private`.")
   }
+
+  def handleDatastoreExceptions[T](f: => T): Try[T] = {
```
This is a little odd. I might use a filter-style function instead:
```scala
/** Take a Try and log any Datastore.Exception encountered. */
def handleDatastore[T](maybeFailure: Try[T]): Try[T] = maybeFailure match {
  case Failure(e: Datastore.Exception) => {
    System.err.println("Error: " + e.getMessage)
    maybeFailure
  }
  case _ => maybeFailure
}
```

dirkgr replied:
To use this, I have to change every function where I want to log the exceptions, and wrap all AWS calls.

dirkgr replied:
Actually, I don't, do I.

dirkgr replied:
But now every place where I use this looks like this:
```
  Common.handleDatastore(Try {
      // do things
  })
```

That's hardly an improvement over this:
```
  Common.handleDatastoreExceptions {
    // do things
  }
```

jkinkead replied:
It's just a suggestion. It's a weird function as-is - you're returning a `Try`, but have an embedded `try` statement, which is odd.

It also only captures `Datastore.DSException`, and will throw any others encountered, which isn't what I'd expect when calling it.

jkinkead commented on datastore-cli/src/main/scala/org/allenai/datastore/cli/Common.scala:
```diff
@@ -1,9 +1,23 @@
+    try {
+      Success(f)
+    } catch {
+      case e: Datastore.Exception =>
```
Braces around body of `case`, per style guide.

dirkgr replied:
IntelliJ says: "Warning: redundant braces"

jkinkead commented on datastore-cli/src/main/scala/org/allenai/datastore/cli/DownloadApp.scala:
```diff
@@ -51,22 +51,24 @@ object DownloadApp extends App {
-    }
+  Common.handleDatastoreExceptions {
+    parser.parse(args, Config()) foreach { config =>
+      val datastore = config.datastore match {
```
```scala
config.datastore getOrElse {
  Common.printDefaultDatastoreWarning()
  Datastore
}
```
?

jkinkead commented on datastore-cli/src/main/scala/org/allenai/datastore/cli/ListApp.scala:
```diff
@@ -18,21 +18,23 @@ object ListApp extends App {
-        }
+  Common.handleDatastoreExceptions {
+    parser.parse(args, Config()) foreach { config =>
+      val datastore = config.datastore match {
```
`getOrElse` here, too?

jkinkead commented on datastore-cli/src/main/scala/org/allenai/datastore/cli/UploadApp.scala:
```diff
@@ -42,20 +42,22 @@ object UploadApp extends App {
-        Datastore
+  Common.handleDatastoreExceptions {
+    parser.parse(args, Config()) foreach { config =>
+      val datastore = config.datastore match {
```
`getOrElse`? Or pull into a helper on `Config`?

Also, you should really rename `Config` - it's confusing when it's not a typesafe config object.

dirkgr replied:
I renamed `Config`, but I'm not very happy about that either. I picked up that naming scheme for scopt from someone else, @schmmd maybe, and this just means that now we have one more scopt naming scheme than we had before.

jkinkead commented on datastore-cli/src/main/scala/org/allenai/datastore/cli/WipeCacheApp.scala:
```diff
@@ -11,13 +11,15 @@ object WipeCacheApp extends App {
-        Datastore
+  Common.handleDatastoreExceptions {
+    parser.parse(args, Config()) foreach { config =>
+      val datastore = config.datastore match {
```
. . . repeat, repeat.

jkinkead commented on datastore/src/main/scala/org/allenai/datastore/Datastore.scala:
```diff
@@ -127,6 +127,9 @@ class Datastore(val name: String, val s3: AmazonS3Client) extends Logging {
   }
 
+  /** Common base class for all datastore exceptions, so they can be caught together */
+  class Exception(message: String, cause: Throwable) extends scala.Exception(message, cause)
```
You shouldn't use a name that conflicts with something in `scala`.

dirkgr replied:
I can see where you're coming from, and I made the change, but it means that to reference this class, you have to say `Datastore.DatastoreException`, which is redundant. When faced with a choice between making something (in this case, the `Datastore` class) nice on the inside, and making it nice on the outside, I always choose the outside. There is a lot more outside than inside.

jkinkead commented:
`Exception` and `Config` should really have non-confusing names.

dirkgr commented:
I put in all the `getOrElse` stuff. It's much better, but it has nothing to do with the change in this PR. It just shows up because all those lines got indented.

dirkgr commented:
I guess what I'm saying is, there might be room to give the whole code a once-over, put braces everywhere you like them, remove parenthesis everywhere IntelliJ doesn't like them, use `Try` instead of exception handling, or maybe the other way around, and so on, but there is marginal utility to opening up and re-arranging code that works and that nobody is touching.

schmmd commented:
Let's amend the style guide to NOT recommend braces on case clauses.  We don't usually do this.

jkinkead commented on datastore/src/main/scala/org/allenai/datastore/Datastore.scala:
```diff
@@ -127,6 +127,9 @@ class Datastore(val name: String, val s3: AmazonS3Client) extends Logging {
   }
 
+  /** Common base class for all datastore exceptions, so they can be caught together */
+  class DSException(message: String, cause: Throwable) extends scala.Exception(message, cause)
```
Technically should be `DsException`.

jkinkead commented:
Keeping your code clean (and inline with the style guide) should be a part of any PR. If you're editing code, you should take the opportunity to clean it up.

@schmmd - The braces are for multi-line case statement. It really helps with readability of longer blocks to have them braced:
```scala
foo match {
  case a: Int => println("it's an int")
  case b: String => {
    doOneThing
    doAnotherThing
    doAThirdThing
  }
}
```

dirkgr commented:
I renamed `DSException`.

dirkgr commented:
If I'm refactoring a piece of code, fixing details along the way makes sense. When all I'm doing is indenting a block, I don't think it does. I don't even necessarily read the whole block in that case. Diff tools other than github will not even show the block.

jkinkead commented:
A comment on `handleDatastoreException` would be very helpful. The fact that it will throw non-DsExceptions (and not capture them in the returned `Try`) is very surprising, and looks like a bug.

jkinkead commented:
My point re: cleanup is that any edits should be taken as opportunities to clean up code. Not that you should read every line you indent, but rather that if you or the reviewer sees a cleanup opportunity, you should jump on it.

dirkgr commented:
I changed the function to return `Unit` instead, and added documentation. The output is never used anyways, and you're right, the signature is confusing.

jkinkead commented:
Awesome, LGTM!
